crates/catalog/sql/src/catalog.rs

```diff
@@ -0,0 +1,397 @@
+                Box::pin(async move {
+                    sqlx::query(
+                        "create table if not exists iceberg_tables (
+                                catalog_name text not null,
+                                table_namespace text not null,
+                                table_name text not null,
+                                metadata_location text not null,
+                                previous_metadata_location text,
+                                primary key (catalog_name, table_namespace, table_name)
+                            );",
+                    )
+                    .execute(&mut **txn)
+                    .await
+                })
+            })
+            .await
+            .map_err(from_sqlx_error)?;
+
+        connection
+            .transaction(|txn| {
+                Box::pin(async move {
+                    sqlx::query(
+                        "create table if not exists iceberg_namespace_properties (
+                                catalog_name text not null,
+                                namespace text not null,
+                                property_key text,
+                                property_value text,
+                                primary key (catalog_name, namespace, property_key)
+                            );",
+                    )
+                    .execute(&mut **txn)
+                    .await
+                })
+            })
+            .await
+            .map_err(from_sqlx_error)?;
+
+        Ok(SqlCatalog {
+            name: name.to_owned(),
+            connection: Arc::new(Mutex::new(connection)),
+            storage,
+            cache: Arc::new(DashMap::new()),
+        })
+    }
+}
+
+#[derive(Debug)]
+struct TableRef {
+    table_namespace: String,
+    table_name: String,
+    metadata_location: String,
+    _previous_metadata_location: Option<String>,
+}
+
+fn query_map(row: &AnyRow) -> std::result::Result<TableRef, sqlx::Error> {
+    Ok(TableRef {
+        table_namespace: row.try_get(0)?,
+        table_name: row.try_get(1)?,
+        metadata_location: row.try_get(2)?,
+        _previous_metadata_location: row.try_get::<String, _>(3).map(Some).or_else(|err| {
+            if let sqlx::Error::ColumnDecode {
+                index: _,
+                source: _,
+            } = err
+            {
+                Ok(None)
+            } else {
+                Err(err)
+            }
+        })?,
+    })
+}
+
+#[async_trait]
+impl Catalog for SqlCatalog {
+    async fn list_namespaces(
+        &self,
+        _parent: Option<&NamespaceIdent>,
+    ) -> Result<Vec<NamespaceIdent>> {
+        let mut connection = self.connection.lock().await;
+        let rows = connection.transaction(|txn|{
+            let name = self.name.clone();
+            Box::pin(async move {
+            sqlx::query(&format!("select distinct table_namespace from iceberg_tables where catalog_name = '{}';",&name)).fetch_all(&mut **txn).await
```

Comment:
Should you care about SQL injections ? Or the catalog / namespace / table names are assumed to be safe ?